{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DrugBank Drug Targets downloaded.\n",
      "ENCODE Transcription Factor Binding Site Profiles downloaded.\n",
      "HPO Gene-Disease Associations downloaded.\n",
      "JASPAR Predicted Transcription Factor Targets downloaded.\n",
      "Reactome Pathways downloaded.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "import requests\n",
    "import os\n",
    "import zlib\n",
    "\n",
    "\n",
    "def _download_file(response, filename):\n",
    "    with open(filename, 'wb') as f:\n",
    "        for chunk in response.iter_content(chunk_size=1024):\n",
    "            f.write(chunk)\n",
    "\n",
    "\n",
    "def _download_and_decompress_file(response, filename):\n",
    "    decompressor = zlib.decompressobj(16 + zlib.MAX_WBITS)\n",
    "    filename = filename[:-3]\n",
    "    with open(filename, 'w+') as f:\n",
    "        while True:\n",
    "            chunk = response.raw.read(1024)\n",
    "            if not chunk:\n",
    "                break\n",
    "            string = decompressor.decompress(chunk)\n",
    "            f.write(string)\n",
    "\n",
    "\n",
    "def download_datasets(selected_datasets, selected_downloads, decompress=False):\n",
    "    for dataset, path in selected_datasets:\n",
    "        if not os.path.exists(dataset):\n",
    "            os.mkdir(dataset)\n",
    "\n",
    "        for downloadable in selected_downloads:\n",
    "            url = 'https://maayanlab.cloud/static/hdfs/harmonizome/data/%s/%s' %\\\n",
    "                  (path, downloadable)\n",
    "            response = requests.get(url, stream=True)\n",
    "            filename = '%s/%s' % (dataset, downloadable)\n",
    "\n",
    "            # Not every dataset has all downloadables.\n",
    "            if response.status_code != 200:\n",
    "                continue\n",
    "\n",
    "            if decompress and 'txt.gz' in filename:\n",
    "                _download_and_decompress_file(response, filename)\n",
    "            else:\n",
    "                _download_file(response, filename)\n",
    "\n",
    "        print('%s downloaded.' % dataset)\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    # Uncomment a dataset to download it.\n",
    "    download_datasets([\n",
    "        # ('Achilles Cell Line Gene Essentiality Profiles', 'achilles'),\n",
    "        # ('Allen Brain Atlas Adult Human Brain Tissue Gene Expression Profiles', 'brainatlasadulthuman'),\n",
    "        # ('Allen Brain Atlas Adult Mouse Brain Tissue Gene Expression Profiles', 'brainatlasadultmouse'),\n",
    "        # ('Allen Brain Atlas Developing Human Brain Tissue Gene Expression Profiles by Microarray', 'brainatlasdevelopmentalhumanmicroarray'),\n",
    "        # ('Allen Brain Atlas Developing Human Brain Tissue Gene Expression Profiles by RNA-seq', 'brainatlasdevelopmentalhumanrnaseq'),\n",
    "        # ('Allen Brain Atlas Prenatal Human Brain Tissue Gene Expression Profiles', 'brainatlasprenatalhuman'),\n",
    "        # ('BIND Biomolecular Interactions', 'bind'),\n",
    "        # ('BioGPS Cell Line Gene Expression Profiles', 'biogpsnci60'),\n",
    "        # ('BioGPS Human Cell Type and Tissue Gene Expression Profiles', 'biogpshuman'),\n",
    "        # ('BioGPS Mouse Cell Type and Tissue Gene Expression Profiles', 'biogpsmouse'),\n",
    "#         ('BioGRID Protein-Protein Interactions', 'biogrid'),\n",
    "        # ('Biocarta Pathways', 'biocarta'),\n",
    "        # ('CCLE Cell Line Gene CNV Profiles', 'cclecnv'),\n",
    "        # ('CCLE Cell Line Gene Expression Profiles', 'cclemrna'),\n",
    "        # ('CCLE Cell Line Gene Mutation Profiles', 'cclemut'),\n",
    "        # ('CHEA Transcription Factor Binding Site Profiles', 'chea'),\n",
    "        # ('CHEA Transcription Factor Targets', 'cheappi'),\n",
    "        # ('CMAP Signatures of Differentially Expressed Genes for Small Molecules', 'cmap'),\n",
    "        # ('COMPARTMENTS Curated Protein Localization Evidence Scores', 'jensencompartmentcurated'),\n",
    "        # ('COMPARTMENTS Experimental Protein Localization Evidence Scores', 'jensencompartmentexpts'),\n",
    "        # ('COMPARTMENTS Text-mining Protein Localization Evidence Scores', 'jensencompartmenttextmining'),\n",
    "        # ('CORUM Protein Complexes', 'corum'),\n",
    "        # ('COSMIC Cell Line Gene CNV Profiles', 'cosmiccnv'),\n",
    "        # ('COSMIC Cell Line Gene Mutation Profiles', 'cosmicmut'),\n",
    "        # ('CTD Gene-Chemical Interactions', 'ctdchemical'),\n",
    "        # ('CTD Gene-Disease Associations', 'ctddisease'),\n",
    "        # ('ClinVar SNP-Phenotype Associations', 'clinvar'),\n",
    "        # ('Combined Pathways Pathways', 'combinedpathways'),\n",
    "        # ('dbGAP Gene-Trait Associations', 'dbgap'),\n",
    "        # ('DEPOD Substrates of Phosphatases', 'depod'),\n",
    "        # ('DIP Protein-Protein Interactions', 'dip'),\n",
    "        # ('DISEASES Curated Gene-Disease Assocation Evidence Scores', 'jensendiseasecurated'),\n",
    "        # ('DISEASES Experimental Gene-Disease Assocation Evidence Scores', 'jensendiseaseexpts'),\n",
    "        # ('DISEASES Text-mining Gene-Disease Assocation Evidence Scores', 'jensendiseasetextmining'),\n",
    "        ('DrugBank Drug Targets', 'drugbank'),\n",
    "        # ('ENCODE Histone Modification Site Profiles', 'encodehm'),\n",
    "        ('ENCODE Transcription Factor Binding Site Profiles', 'encodetf'),\n",
    "        # ('ENCODE Transcription Factor Targets', 'encodetfppi'),\n",
    "        # ('ESCAPE Omics Signatures of Genes and Proteins for Stem Cells', 'escape'),\n",
    "        # ('GAD Gene-Disease Associations', 'gad'),\n",
    "        # ('GAD High Level Gene-Disease Associations', 'gadhighlevel'),\n",
    "        # ('GDSC Cell Line Gene Expression Profiles', 'gdsc'),\n",
    "        # ('GEO Signatures of Differentially Expressed Genes for Diseases', 'geodisease'),\n",
    "        # ('GEO Signatures of Differentially Expressed Genes for Gene Perturbations', 'geogene'),\n",
    "        # ('GEO Signatures of Differentially Expressed Genes for Kinase Perturbations', 'geokinase'),\n",
    "        # ('GEO Signatures of Differentially Expressed Genes for Small Molecules', 'geochemical'),\n",
    "        # ('GEO Signatures of Differentially Expressed Genes for Transcription Factor Perturbations', 'geotf'),\n",
    "        # ('GEO Signatures of Differentially Expressed Genes for Viral Infections', 'geovirus'),\n",
    "        # ('GO Biological Process Annotations', 'gobp'),\n",
    "        # ('GO Cellular Component Annotations', 'gocc'),\n",
    "        # ('GO Molecular Function Annotations', 'gomf'),\n",
    "        # ('GTEx Tissue Gene Expression Profiles', 'gtextissue'),\n",
    "        # ('GTEx Tissue Sample Gene Expression Profiles', 'gtexsample'),\n",
    "        # ('GTEx eQTL', 'gtexeqtl'),\n",
    "        # ('GWAS Catalog SNP-Phenotype Associations', 'gwascatalog'),\n",
    "        # ('GWASdb SNP-Disease Associations', 'gwasdbdisease'),\n",
    "        # ('GWASdb SNP-Phenotype Associations', 'gwasdbphenotype'),\n",
    "        # ('GeneRIF Biological Term Annotations', 'generif'),\n",
    "        # ('GeneSigDB Published Gene Signatures', 'genesigdb'),\n",
    "        # ('Graph of Medicine EHR Text-mining Clinical Term Annotations', 'graphofmedicine'),\n",
    "        # ('Guide to Pharmacology Chemical Ligands of Receptors', 'guidetopharmchemical'),\n",
    "        # ('Guide to Pharmacology Protein Ligands of Receptors', 'guidetopharmprotein'),\n",
    "        # ('HMDB Metabolites of Enzymes', 'hmdb'),\n",
    "        # ('HPA Cell Line Gene Expression Profiles', 'hpacelllines'),\n",
    "        # ('HPA Tissue Gene Expression Profiles', 'hpatissuesmrna'),\n",
    "        # ('HPA Tissue Protein Expression Profiles', 'hpatissuesprotein'),\n",
    "        # ('HPA Tissue Sample Gene Expression Profiles', 'hpasamples'),\n",
    "        # ('HPM Cell Type and Tissue Protein Expression Profiles', 'hpm'),\n",
    "        ('HPO Gene-Disease Associations', 'hpo'),\n",
    "        # ('HPRD Protein-Protein Interactions', 'hprd'),\n",
    "        # ('Heiser et al., PNAS, 2011 Cell Line Gene Expression Profiles', 'heiser'),\n",
    "        # ('HuGE Navigator Gene-Phenotype Associations', 'hugenavigator'),\n",
    "        # ('Hub Proteins Protein-Protein Interactions', 'hubs'),\n",
    "        # ('HumanCyc Biomolecular Interactions', 'humancycppi'),\n",
    "        # ('HumanCyc Pathways', 'humancyc'),\n",
    "        # ('IntAct Biomolecular Interactions', 'intact'),\n",
    "        # ('InterPro Predicted Protein Domain Annotations', 'interpro'),\n",
    "        ('JASPAR Predicted Transcription Factor Targets', 'jasparpwm'),\n",
    "        # ('KEA Substrates of Kinases', 'kea'),\n",
    "        # ('KEGG Biomolecular Interactions', 'keggppi'),\n",
    "        # ('KEGG Pathways', 'kegg'),\n",
    "        # ('Kinativ Kinase Inhibitor Bioactivity Profiles', 'kinativ'),\n",
    "        # ('KinomeScan Kinase Inhibitor Targets', 'kinomescan'),\n",
    "        # ('Klijn et al., Nat. Biotechnol., 2015 Cell Line Gene CNV Profiles', 'klijncnv'),\n",
    "        # ('Klijn et al., Nat. Biotechnol., 2015 Cell Line Gene Expression Profiles', 'klijnmrna'),\n",
    "        # ('Klijn et al., Nat. Biotechnol., 2015 Cell Line Gene Mutation Profiles', 'klijnmut'),\n",
    "        # ('LINCS L1000 CMAP Signatures of Differentially Expressed Genes for Gene Knockdowns', 'lincscmapgene'),\n",
    "        # ('LINCS L1000 CMAP Signatures of Differentially Expressed Genes for Small Molecules', 'lincscmapchemical'),\n",
    "        # ('LOCATE Curated Protein Localization Annotations', 'locate'),\n",
    "        # ('LOCATE Predicted Protein Localization Annotations', 'locatepredicted'),\n",
    "        # ('MPO Gene-Phenotype Associations', 'mgimpo'),\n",
    "        # ('MSigDB Cancer Gene Co-expression Modules', 'msigdbcomp'),\n",
    "        # ('MSigDB Signatures of Differentially Expressed Genes for Cancer Gene Perturbations', 'msigdbonc'),\n",
    "        # ('MiRTarBase microRNA Targets', 'mirtarbase'),\n",
    "        # ('MotifMap Predicted Transcription Factor Targets', 'motifmap'),\n",
    "        # ('NURSA Protein Complexes', 'nursa'),\n",
    "        # ('NURSA Protein-Protein Interactions', 'nursappi'),\n",
    "        # ('OMIM Gene-Disease Associations', 'omim'),\n",
    "        # ('PANTHER Biomolecular Interactions', 'pantherppi'),\n",
    "        # ('PANTHER Pathways', 'panther'),\n",
    "        # ('PID Biomolecular Interactions', 'pidppi'),\n",
    "        # ('PID Pathways', 'pid'),\n",
    "        # ('Pathway Commons Protein-Protein Interactions', 'pc'),\n",
    "        # ('PhosphoSitePlus Phosphosite-Disease Associations', 'phosphositeplusdisease'),\n",
    "        # ('PhosphoSitePlus Substrates of Kinases', 'phosphositeplus'),\n",
    "        # ('Phosphosite Textmining Biological Term Annotations', 'phosphositetextmining'),\n",
    "        # ('ProteomicsDB Cell Type and Tissue Protein Expression Profiles', 'proteomicsdb'),\n",
    "        # ('Reactome Biomolecular Interactions', 'reactomeppi'),\n",
    "        ('Reactome Pathways', 'reactome'),\n",
    "        # ('Recon X Predicted Biomolecular Interactions', 'reconx'),\n",
    "        # ('Roadmap Epigenomics Cell and Tissue DNA Accessibility Profiles', 'epigenomicsdnaaccessibility'),\n",
    "        # ('Roadmap Epigenomics Cell and Tissue DNA Methylation Profiles', 'epigenomicsdnamethylation'),\n",
    "        # ('Roadmap Epigenomics Cell and Tissue Gene Expression Profiles', 'epigenomicsmrna'),\n",
    "        # ('Roadmap Epigenomics Histone Modification Site Profiles', 'epigenomicshm'),\n",
    "        # ('SILAC Phosphoproteomics Signatures of Differentially Phosphorylated Proteins for Drugs', 'silacdrug'),\n",
    "        # ('SILAC Phosphoproteomics Signatures of Differentially Phosphorylated Proteins for Gene Perturbations', 'silacgene'),\n",
    "        # ('SILAC Phosphoproteomics Signatures of Differentially Phosphorylated Proteins for Protein Ligands', 'silacligand'),\n",
    "        # ('SNPedia SNP-Phenotype Associations', 'snpedia'),\n",
    "        # ('TCGA Signatures of Differentially Expressed Genes for Tumors', 'tcga'),\n",
    "        # ('TISSUES Curated Tissue Protein Expression Evidence Scores', 'jensentissuecurated'),\n",
    "        # ('TISSUES Experimental Tissue Protein Expression Evidence Scores', 'jensentissueexpts'),\n",
    "        # ('TISSUES Text-mining Tissue Protein Expression Evidence Scores', 'jensentissuetextmining'),\n",
    "        # ('TRANSFAC Curated Transcription Factor Targets', 'transfac'),\n",
    "        # ('TRANSFAC Predicted Transcription Factor Targets', 'transfacpwm'),\n",
    "        # ('TargetScan Predicted Conserved microRNA Targets', 'targetscan'),\n",
    "        # ('TargetScan Predicted Nonconserved microRNA Targets', 'targetscannonconserved'),\n",
    "        # ('Virus MINT Protein-Viral Protein Interactions', 'virusmintppi'),\n",
    "        # ('Virus MINT Protein-Virus Interactions', 'virusmint'),\n",
    "        # ('Wikipathways Pathways', 'wikipathways'),\n",
    "    ], [\n",
    "        # 'gene_attribute_matrix.txt.gz',\n",
    "        # 'gene_attribute_edges.txt.gz',\n",
    "        # 'gene_set_library_crisp.txt.gz',\n",
    "        # 'gene_set_library_up_crisp.txt.gz',\n",
    "        # 'gene_set_library_dn_crisp.txt.gz',\n",
    "        # 'attribute_set_library_crisp.txt.gz',\n",
    "        # 'attribute_set_library_up_crisp.txt.gz',\n",
    "        # 'attribute_set_library_dn_crisp.txt.gz',\n",
    "        'gene_similarity_matrix_cosine.txt.gz',\n",
    "        # 'attribute_similarity_matrix_cosine.txt.gz',\n",
    "        # 'gene_list_terms.txt.gz',\n",
    "        # 'attribute_list_entries.txt.gz',\n",
    "        # 'processing_script.m'\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
